src/torch/torch.py

```diff
@@ -0,0 +1,517 @@
+try:
+    import torch
+except ModuleNotFoundError:
+    raise ModuleNotFoundError('Failed to import torch. Please install PyTorch first.')
+
+
+import sirf
+import numpy
+
+
+
+# This module provides a PyTorch interface for SIRF operators and objective functions.
+# It wraps SIRF objects to enable their use within PyTorch's autograd system.
+# The core idea is to use torch.autograd.Function to define custom forward and backward
+# passes that interact with SIRF's C++ backend.  This allows SIRF operations to be
+# seamlessly integrated into PyTorch computational graphs.
+#
+# based on 
+# https://github.com/educating-dip/pet_deep_image_prior/blob/main/src/deep_image_prior/torch_wrapper.py
+
+
+def sirf_to_torch(
+        sirf_src: sirf.DataContainer | float,
+        device: torch.device,
+        requires_grad: bool = False
+        ) -> torch.Tensor:
+    """
+    Converts a SIRF object to a PyTorch tensor.
+
+    Args:
+        sirf_src: The SIRF object to convert.  This can be a SIRF 
+          `DataContainer` such as `AcquisitionData`, `ImageData`, or a scalar (float).
+        device: The PyTorch device to place the tensor on (e.g., 'cpu' or 
+          'cuda').
+        requires_grad:  Whether the resulting tensor should track gradients.
+
+    Returns:
+        A PyTorch tensor representing the SIRF object's data.
+
+    Raises:
+        TypeError: If `sirf_src` is not a supported SIRF object or a float.
+    """
+
+    if isinstance(sirf_src, float):
+        # Handle float separately for efficiency and type consistency.
+        return torch.tensor(sirf_src, requires_grad=requires_grad, \
+            device=device
+            )
+    elif hasattr(sirf_src, 'as_array'):  
+        # Check for as_array method (ImageData, AcquisitionData, etc.)
+        return torch.tensor(sirf_src.as_array(), requires_grad=requires_grad,
+            device=device
+            )
+    else:
+        raise TypeError(f"Unsupported SIRF object type: {type(sirf_src)}")
+
+def torch_to_sirf_(
```

Comment:
any reason this has an underscore suffix?

Reply:
suggested by @ckolbPTB to follow torch conventions. I cannot remember why. something to do with autograd. Probably a good idea to document

Reply:
It's because there is an in-place operation that can be dangerous in PyTorch (breaks the computational graph). Trailing underscore indicates from what I understand!